presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpFilterToKqlConverter.java

```diff
@@ -0,0 +1,679 @@
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.RowExpressionVisitor;
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.google.common.collect.ImmutableSet;
+import io.airlift.slice.Slice;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.common.function.OperatorType.EQUAL;
+import static com.facebook.presto.common.function.OperatorType.GREATER_THAN;
+import static com.facebook.presto.common.function.OperatorType.GREATER_THAN_OR_EQUAL;
+import static com.facebook.presto.common.function.OperatorType.LESS_THAN;
+import static com.facebook.presto.common.function.OperatorType.LESS_THAN_OR_EQUAL;
+import static com.facebook.presto.common.function.OperatorType.NOT_EQUAL;
+import static com.facebook.presto.common.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.plugin.clp.ClpErrorCode.CLP_PUSHDOWN_UNSUPPORTED_EXPRESSION;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.AND;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * ClpFilterToKqlConverter translates Presto RowExpressions into KQL (Kibana Query Language) filters
+ * used as CLP queries. This is used primarily for pushing down supported filters to the CLP engine.
+ * This class implements the RowExpressionVisitor interface and recursively walks Presto filter expressions,
+ * attempting to convert supported expressions (e.g., comparisons, logical AND/OR, LIKE, IN, IS NULL,
+ * and SUBSTR-based expressions) into corresponding KQL filter strings. Any part of the expression that
+ * cannot be translated is preserved as a "remaining expression" for potential fallback processing.
+ * Supported translations include:
+ * - Variable-to-literal comparisons (e.g., =, !=, <, >, <=, >=)
+ * - String pattern matches using LIKE
+ * - Membership checks using IN
+ * - NULL checks via IS NULL
+ * - Substring comparisons (e.g., SUBSTR(x, start, len) = "val") mapped to wildcard KQL queries
+ * - Dereferencing fields from row-typed variables
+ * - Logical operators AND, OR, and NOT
+ */
+public class ClpFilterToKqlConverter
+        implements RowExpressionVisitor<ClpExpression, Void>
+{
+    private static final Set<OperatorType> LOGICAL_BINARY_OPS_FILTER =
+            ImmutableSet.of(EQUAL, NOT_EQUAL, LESS_THAN, LESS_THAN_OR_EQUAL, GREATER_THAN, GREATER_THAN_OR_EQUAL);
+
+    private final StandardFunctionResolution standardFunctionResolution;
+    private final FunctionMetadataManager functionMetadataManager;
+    private final Map<VariableReferenceExpression, ColumnHandle> assignments;
+
+    public ClpFilterToKqlConverter(StandardFunctionResolution standardFunctionResolution,
+            FunctionMetadataManager functionMetadataManager,
+            Map<VariableReferenceExpression, ColumnHandle> assignments)
```

Comment:
by CS (code-style), the first argument cannot be in the same line with the method name if you put the rest arguments each in a separated line.
```suggestion
    public ClpFilterToKqlConverter(
            StandardFunctionResolution standardFunctionResolution,
            FunctionMetadataManager functionMetadataManager,
            Map<VariableReferenceExpression, ColumnHandle> assignments)
```